### DISCUSSION TITLE
How to keep diff view open?

### DISCUSSION THREAD

**Discussion by ChrisTaylorDeveloper** [sourcegit-scm/sourcegit#754] [Q&A] [answered]
Is it possible to keep the diff view open as I move up and down the commits in the graph?

If I select a commit in the graph, then select CHANGES, then select a file I see the diff view.  However,  if I then move to another commit, the diff view closes until I select a changed file again.

If there are multiple changed files per commit then showing the diff for the first file in the list would be fine.

**Comment by love-linger** [comment, marked as answer]
This issue has been raised many times.

The mainstream suggestions currently received include:

* After the repository is loaded, the current `HEAD` should be selected automatically.
* Always switch to the `INFORMATION` page after commit changed to quickly find out if this commit is the one we want to get.
* Keep the current page, and if it is not the `INFORMATION` page, there are two different suggestions
  - Automatically select the first file or change
  - Automatically select the last viewed change or file

The main reasons why `SourceGit` did not choose one of these options are:

* As can be seen from the above, each solution has its own specific situation.
* When the viewing commit changes, a series of `Git` commands are executed, including but not limited to
  - get the entire message of this commit
  - get the change file list included in current commit
  - get the files that contains in the revision of the commit

  and after adding automatic selection, it will increase again.
  - calcuate the diff of the selected change
  - get the entire content of the selected revision file
  
  on some low-configured machines, too many git process will stuck the app.
* For some users or at certain times, the results of these auto-selection are meaningless. For example, suppose that you are switching commits quickly and frequently to find the one you need. 
* I don't think it's difficult for users to manually select the files they want to view when they need to.

**Reply by ChrisTaylorDeveloper** [reply]
Thank you for your detailed reply.

**Comment by goran-w** [comment]
I requested this too right now, in #1019, before I was aware of this discussion.

> This issue has been raised many times.

That might indicate that it's something many would like to see implemented...

It's also how it works in many other tools: Plastic SCM, GitHub Desktop, Fork, Git Extensions, Tower, Sourcetree etc.

> on some low-configured machines, too many git process will stuck the app.

This should be implemented as an optional feature (by Preferences) - not everyone has a slow machine.

> suppose that you are switching commits quickly and frequently to find the one you need.

Suppose that you have several commits in a row, changing a single file, or a few related ones. Having auto-selection would avoid several unnecessary clicks of having to re-select when quickly and frequently switching commits to find the one you need.

**Reply by love-linger** [reply]
> Suppose that you have several commits in a row, changing a single file, or a few related ones. Having auto-selection would avoid several unnecessary clicks of having to re-select when quickly and frequently switching commits to find the one you need.

I suggest you use the `File Histories` function. Suppose that there's one commit that does not contains the modification of the target file, or the file is not always the first one in the tree/list - In this case, the automatically selected feature will not run as expected, and no helps.

**Reply by goran-w** [reply]
Agreed, if looking at a single file, but often I'm looking at work on a branch where a small set of related files are changed, and having the first file auto-selected would still help a lot with making the navigation quicker and keeping the changes fresh in memory. Since I'm used to this behavior in previous tools, the current situation in SourceGit is a downgrade.

**Reply by goran-w** [reply]
> As can be seen from the above, each solution has its own specific situation.

This should not be taken as a reason to dismiss all of them in one go.

Instead, these solutions could thereby be implemented separately, and each/any of them would be an improvement.

**Reply by goran-w** [reply]
> For some users or at certain times, the results of these auto-selection are meaningless.

That's why it should be an optional feature - then each user can choose their own trade-off between convenience and performance...

**Reply by goran-w** [reply]
> Keep the current page, and if it is not the INFORMATION page, there are two different suggestions
• Automatically select the first file or change
• Automatically select the last viewed change or file

The 2nd of these two suggestions (select last viewed) seems unnecessarily complicated, since it would need more bookkeeping.
The 1st suggestion (select first file-change) is much simpler and should be a sufficient improvement.

**Reply by goran-w** [reply]
> * After the repository is loaded, the current HEAD should be selected automatically.

This is less important IMHO, as it would only happen once on SourceGit startup (and when switching to a repo tab first time since startup). Still nice to have, though.

BTW, since LOCAL BRANCHES are expanded by default, if there are a LOT of local branches it would be nice to have the active (HEAD-pointed) branch scrolled into view whenever a repo tab becomes activated (switched-to)... Reported as suggestion in #1022.

> * Always switch to the INFORMATION page after commit changed to quickly find out if this commit is the one we want to get.

No, please! I want the INFORMATION / CHANGES / FILES tab to stay as I last selected it. It would be very annoying if it kept resetting (as it currently does when switching back to single-selection from multi-selection, which I point out in #1019).

**Reply by love-linger** [reply]
I've push some changes. You can test the latest CI build from Github Action.

**Reply by goran-w** [reply]
> I've push some changes. You can test the latest CI build from Github Action.

Thanks, this is great! (For details, see commits referenced in #1019.)
